libafl/src/inputs/multi.rs

```diff
@@ -0,0 +1,165 @@
+//! Definitions for inputs which have multiple distinct subcomponents.
+//!
+//! Unfortunately, since both [`serde::de::Deserialize`] and [`Clone`] require [`Sized`], it is not
+//! possible to dynamically define a single input with dynamic typing. As such, [`MultipartInput`]
+//! requires that each subcomponent be the same subtype.
+
+use alloc::{
+    string::{String, ToString},
+    vec::Vec,
+};
+
+use arrayvec::ArrayVec;
+use serde::{Deserialize, Serialize};
+
+use crate::inputs::Input;
+
+/// An input composed of multiple parts. Use in situations where subcomponents are not necessarily
+/// related, or represent distinct parts of the input.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MultipartInput<I> {
+    parts: Vec<I>,
+    names: Vec<String>,
+}
+
+impl<I> Default for MultipartInput<I> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<I> MultipartInput<I> {
+    /// Create a new multipart input.
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            parts: Vec::new(),
+            names: Vec::new(),
+        }
+    }
+
+    fn idxs_to_skips(idxs: &mut [usize]) {
+        for following in (1..idxs.len()).rev() {
+            let first = idxs[following - 1];
+            let second = idxs[following];
+
+            idxs[following] = second
+                .checked_sub(first)
+                .expect("idxs was not sorted")
+                .checked_sub(1)
+                .expect("idxs had duplicate elements");
+        }
+    }
+
+    /// Get the individual parts of this input.
+    #[must_use]
+    pub fn parts(&self) -> &[I] {
+        &self.parts
+    }
+
+    /// Access multiple parts mutably.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if idxs is not sorted, has duplicate elements, or any entry is out of bounds.
+    #[must_use]
+    pub fn parts_mut<const N: usize>(&mut self, mut idxs: [usize; N]) -> [&mut I; N] {
```

Comment:
Would it make sense to have a special case for bytesinput that can create sub-inputs from a bytesinput on the fly?

Reply:
Unless the bytesinputs are well-formed and a strong transformation rule (e.g. JSON?) is implemented, this is going to be implementation-defined.